### PR TITLE
Refactor Address input component test file to new format

### PR DESCRIPTION
**Pull request by precious-onyenaucheya-ons**
### What is the context of this PR?

Fixes: ONSdigital/design-team#132
This PR involves refactoring of the  address input test document to follow the Given, When, Then testing approach.

### How to review this PR

**Functionality:**
- Run the test suite using: yarn test --testNamePattern="FOR: Macro: Address-input"

**Does this testing refactor meet the following success criteria:**
- Clearly structured.
- Follows naming conventions.
- Consistent.
- Concise.
- Suited to our codebase.
- Compatible with existing testing framework.

**Look at the refactored code in relation to:**
- The original test document.
- The macro file.
- The macro options file.

### Checklist

This needs to be completed by the person raising the PR.

- [x] I have selected the correct Assignee
- [x] I have linked the correct Issue

### CODE REVIEW

**Comment by rmccar** [comment]
I think part of the work on this was to move the examples like EXAMPLE_MANUAL_INPUT_FIELDS out into its own file and then be imported. Have a look at one of the ones Bali has done

File: src/components/address-input/autosuggest.address.spec.js [hunk 1]
```diff
@@ -180,103 +120,218 @@ describe('script: address-input', () => {
     beforeEach(async () => {
         await apiFaker.reset();
     });
+    describe('GIVEN: the component initializes', () => {
+        describe('WHEN: the component initializes without options', () => {
```
Comment (rmccar) [inline src/components/address-input/autosuggest.address.spec.js:124]: Theres some repetition here. Im not sure if we have set rules around what should go in GIVEN and WHENs in script tests. This may need some discussion.

File: src/components/address-input/_macro.spec.js [hunk 2]
```diff
@@ -431,10 +361,12 @@ describe('macro: address-input', () => {
                 dontWrap: true,
             });
 
-            expect(fieldsetSpy.occurrences.length).toBe(0);
+            test('THEN: it does not render `fieldset` component`', () => {
```
Comment (rmccar) [inline src/components/address-input/_macro.spec.js:364]:
```suggestion
            test('THEN: it does not render fieldset component', () => {
```

File: src/components/address-input/_macro.spec.js [hunk 3]
```diff
@@ -416,13 +344,15 @@ describe('macro: address-input', () => {
...
+                expect($('.ons-js-address-manual-btn').attr('href')).toBe('https://example.com/edit-address');
+                expect($('.ons-js-address-manual-btn').text().trim()).toBe('Manually enter address');
+            });
         });
     });
 
-    describe('fieldset', () => {
-        it('does not render `fieldset` component when `dontWrap` is `true`', () => {
+    describe('GIVEN: fieldset', () => {
+        describe('WHEN: `dontWrap` is `true`', () => {
```
Comment (rmccar) [inline src/components/address-input/_macro.spec.js:355]:
```suggestion
        describe('WHEN: dontWrap is true', () => {
```

File: src/components/address-input/_macro.spec.js [hunk 4]
```diff
@@ -431,10 +361,12 @@ describe('macro: address-input', () => {
...
             });
 
-            expect(fieldsetSpy.occurrences.length).toBe(0);
+            test('THEN: it does not render `fieldset` component`', () => {
+                expect(fieldsetSpy.occurrences.length).toBe(0);
+            });
         });
 
-        it('renders `fieldset` component with expected parameters', () => {
+        describe('WHEN: `dontWrap` is not set', () => {
```
Comment (rmccar) [inline src/components/address-input/_macro.spec.js:369]:
```suggestion
        describe('WHEN: dontWrap is not set', () => {
```

File: src/components/address-input/_macro.spec.js [hunk 5]
```diff
@@ -444,11 +376,13 @@ describe('macro: address-input', () => {
                 legendClasses: 'extra-legend-class',
             });
 
-            expect(fieldsetSpy.occurrences[0]).toEqual({
-                id: 'address-input-example-id',
-                classes: 'extra-field-class',
-                legend: 'What is the address?',
-                legendClasses: 'extra-legend-class',
+            test('THEN: it renders `fieldset` component with expected parameters', () => {
```
Comment (rmccar) [inline src/components/address-input/_macro.spec.js:379]:
```suggestion
            test('THEN: it renders the fieldset component with expected parameters', () => {
```

File: src/components/address-input/_macro.spec.js [hunk 6]
```diff
@@ -416,13 +344,15 @@ describe('macro: address-input', () => {
                 }),
             );
 
-            expect($('.ons-js-address-manual-btn').attr('href')).toBe('https://example.com/edit-address');
-            expect($('.ons-js-address-manual-btn').text().trim()).toBe('Manually enter address');
+            test('THEN: renders `manualLinkText` with `manualLink`', () => {
```
Comment (rmccar) [inline src/components/address-input/_macro.spec.js:347]: I think the wording of this test could be more descriptive on what it is testing

File: src/components/address-input/_macro.spec.js [hunk 7]
```diff
@@ -345,69 +269,73 @@ describe('macro: address-input', () => {
...
-            expect($('.ons-js-address-manual-btn').attr('href')).toBe('#0');
-            expect($('.ons-js-address-manual-btn').text().trim()).toBe('Manually enter address');
+            test('THEN: it renders manualLinkText', () => {
+                expect($('.ons-js-address-manual-btn').attr('href')).toBe('#0');
+                expect($('.ons-js-address-manual-btn').text().trim()).toBe('Manually enter address');
+            });
         });
 
-        it('renders `manualLinkText` with `manualLink` when provided', () => {
+        describe('WHEN: a value for `manualLink` is provided', () => {
```
Comment (rmccar) [inline src/components/address-input/_macro.spec.js:338]:
```suggestion
        describe('WHEN: a value for manualLink is provided', () => {
```

File: src/components/address-input/_macro.spec.js [hunk 8]
```diff
@@ -345,69 +269,73 @@ describe('macro: address-input', () => {
...
             const $ = cheerio.load(
                 renderComponent('address-input', {
                     ...EXAMPLE_AUTOSUGGEST_ADDRESS_MINIMAL,
                     manualLinkText: 'Manually enter address',
                 }),
             );
 
-            expect($('.ons-js-address-manual-btn').attr('href')).toBe('#0');
-            expect($('.ons-js-address-manual-btn').text().trim()).toBe('Manually enter address');
+            test('THEN: it renders manualLinkText', () => {
```
Comment (rmccar) [inline src/components/address-input/_macro.spec.js:332]: This could have a better description

File: src/components/address-input/_macro.spec.js [hunk 9]
```diff
@@ -345,69 +269,73 @@ describe('macro: address-input', () => {
...
+                    errorMessageAPILinkText: '[params.errorMessageAPILinkText]',
+                    options: '[params.options]',
+                    manualLink: '[params.manualLink]',
+                    manualLinkText: '[params.manualLinkText]',
+                });
             });
         });
 
-        it('renders manualLinkText` when provided with a default value for `manualLink`', () => {
+        describe('WHEN: provided with a default value for `manualLink`', () => {
```
Comment (rmccar) [inline src/components/address-input/_macro.spec.js:324]:
```suggestion
        describe('WHEN: provided with a default value for manualLink', () => {
```

File: src/components/address-input/_macro.spec.js [hunk 10]
```diff
@@ -291,10 +213,12 @@ describe('macro: address-input', () => {
...
             });
 
-            expect(autosuggestSpy.occurrences.length).toBe(0);
+            test('THEN: it is not shown', () => {
+                expect(autosuggestSpy.occurrences.length).toBe(0);
+            });
         });
 
-        it('renders `autosuggest` component with expected parameters', () => {
+        describe('WHEN: `manualEntry` is not set', () => {
```
Comment (rmccar) [inline src/components/address-input/_macro.spec.js:221]:
```suggestion
        describe('WHEN: manualEntry is not set', () => {
```

File: src/components/address-input/_macro.spec.js [hunk 11]
```diff
@@ -271,18 +191,20 @@ describe('macro: address-input', () => {
...
+                    value: '[params.uprn.value]',
+                });
             });
         });
     });
 
-    describe('autosuggest search field', () => {
-        it('is not shown when `manualEntry` is `true`', () => {
+    describe('GIVEN: autosuggest search field', () => {
+        describe('WHEN: `manualEntry` is `true`', () => {
```
Comment (rmccar) [inline src/components/address-input/_macro.spec.js:207]:
```suggestion
        describe('WHEN: manualEntry is true', () => {
```

File: src/components/address-input/_macro.spec.js [hunk 12]
```diff
@@ -271,18 +191,20 @@ describe('macro: address-input', () => {
                 },
             });
 
-            expect(inputSpy.occurrences).toContainEqual({
-                id: 'address-input-example-id-uprn',
-                classes: 'ons-js-hidden-uprn ons-u-d-no',
-                type: 'hidden',
-                name: 'address-input-example-id-uprn',
-                value: '[params.uprn.value]',
+            test('THEN: it renders hidden `input` component with expected parameters', () => {
```
Comment (rmccar) [inline src/components/address-input/_macro.spec.js:194]:
```suggestion
            test('THEN: it renders hidden input component with expected parameters', () => {
```

File: src/components/address-input/_macro.spec.js [hunk 13]
```diff
@@ -228,39 +152,35 @@ describe('macro: address-input', () => {
...
+                    classes: 'ons-js-hidden-uprn ons-u-d-no',
+                    type: 'hidden',
+                    name: 'address-input-example-id-uprn',
+                    value: '',
+                });
             });
         });
 
-        it('renders hidden `input` component with expected parameters when `uprn.value` is provided', () => {
+        describe('WHEN: `uprn.value` is provided', () => {
```
Comment (rmccar) [inline src/components/address-input/_macro.spec.js:183]:
```suggestion
        describe('WHEN: uprn.value is provided', () => {
```

File: src/components/address-input/_macro.spec.js [hunk 14]
```diff
@@ -228,39 +152,35 @@ describe('macro: address-input', () => {
...
 
             faker.renderComponent('address-input', EXAMPLE_AUTOSUGGEST_ADDRESS_MINIMAL);
 
-            expect(inputSpy.occurrences).toContainEqual({
-                id: 'address-input-example-id-uprn',
-                classes: 'ons-js-hidden-uprn ons-u-d-no',
-                type: 'hidden',
-                name: 'address-input-example-id-uprn',
-                value: '',
+            test('THEN: it renders hidden `input` component with expected parameters', () => {
```
Comment (rmccar) [inline src/components/address-input/_macro.spec.js:172]:
```suggestion
            test('THEN: it renders hidden input component with expected parameters', () => {
```

File: src/components/address-input/_macro.spec.js [hunk 15]
```diff
@@ -228,39 +152,35 @@ describe('macro: address-input', () => {
...
+            test('THEN: it renders provided text for search button', () => {
+                expect($('.ons-js-address-search-btn').text().trim()).toBe('Search for address');
+            });
         });
     });
 
-    describe('hidden field for uprn', () => {
-        it('renders hidden `input` component with expected parameters when `uprn.value` is not provided', () => {
+    describe('GIVEN: hidden field for uprn', () => {
+        describe('WHEN: `uprn.value` is not provided', () => {
```
Comment (rmccar) [inline src/components/address-input/_macro.spec.js:166]:
```suggestion
        describe('WHEN: uprn.value is not provided', () => {
```

File: src/components/address-input/_macro.spec.js [hunk 16]
```diff
@@ -215,11 +138,12 @@ describe('macro: address-input', () => {
...
             );
-
-            expect($('.ons-js-address-search-btn').length).toBe(0);
+            test('THEN: it does not render', () => {
+                expect($('.ons-js-address-search-btn').length).toBe(0);
+            });
         });
 
-        it('marks field so that it is displayed only when there is no javascript', () => {
+        describe('WHEN: there is no javascript', () => {
```
Comment (rmccar) [inline src/components/address-input/_macro.spec.js:146]: This should be something like "WHEN: javascript is turned off in the browser"

File: src/components/address-input/_macro.spec.js [hunk 17]
```diff
@@ -228,39 +152,35 @@ describe('macro: address-input', () => {
...
-
-        it('renders provided text for search button', () => {
-            const $ = cheerio.load(
-                renderComponent('address-input', {
-                    ...EXAMPLE_AUTOSUGGEST_ADDRESS_MINIMAL,
-                    isEditable: true,
-                    searchButton: 'Search for address',
-                }),
-            );
+            test('THEN: it marks field so that it is displayed', () => {
```
Comment (rmccar) [inline src/components/address-input/_macro.spec.js:155]: This description could better describe what is being tested

File: src/components/address-input/_macro.spec.js [hunk 18]
```diff
@@ -228,39 +152,35 @@ describe('macro: address-input', () => {
...
+                expect($('.ons-js-address-search-btn').hasClass('ons-u-db-no-js_disabled')).toBe(true);
+            });
 
-            expect($('.ons-js-address-search-btn').text().trim()).toBe('Search for address');
+            test('THEN: it renders provided text for search button', () => {
```
Comment (rmccar) [inline src/components/address-input/_macro.spec.js:159]:
```suggestion
            test('THEN: it renders search button with provided text', () => {
```

File: src/components/address-input/_macro.spec.js [hunk 19]
```diff
@@ -215,11 +138,12 @@ describe('macro: address-input', () => {
                     searchButton: 'Search for address',
                 }),
             );
-
-            expect($('.ons-js-address-search-btn').length).toBe(0);
+            test('THEN: it does not render', () => {
```
Comment (rmccar) [inline src/components/address-input/_macro.spec.js:141]: Should say what isnt rendered

File: src/components/address-input/_macro.spec.js [hunk 20]
```diff
@@ -87,56 +42,12 @@ describe('macro: address-input', () => {
...
-                id: 'address-input-example-id-line1',
-                name: 'address-input-example-id-line1',
-                classes: 'ons-js-address-line1',
-                label: {
-                    text: 'Address line 1',
-                },
-                value: 'Flat 12345',
-                width: '20@m',
-                error: { text: 'Server error: address line 1' },
+            test('THEN: does not have class to hide input fields', () => {
```
Comment (rmccar) [inline src/components/address-input/_macro.spec.js:45]:
```suggestion
            test('THEN: it does not have class to hide input fields', () => {
```

File: src/components/address-input/_macro.spec.js [hunk 21]
```diff
@@ -416,13 +344,15 @@ describe('macro: address-input', () => {
                 }),
             );
 
-            expect($('.ons-js-address-manual-btn').attr('href')).toBe('https://example.com/edit-address');
-            expect($('.ons-js-address-manual-btn').text().trim()).toBe('Manually enter address');
+            test('THEN: it renders the provided manualLink with the manualText', () => {
```
Comment (rmccar) [inline src/components/address-input/_macro.spec.js:347]:
```suggestion
            test('THEN: it renders the manual link with the provided text and url, () => {
```

File: src/components/address-input/_macro.spec.js [hunk 22]
```diff
@@ -215,11 +138,12 @@ describe('macro: address-input', () => {
...
             );
-
-            expect($('.ons-js-address-search-btn').length).toBe(0);
+            test('THEN: search button does not render', () => {
+                expect($('.ons-js-address-search-btn').length).toBe(0);
+            });
         });
 
-        it('marks field so that it is displayed only when there is no javascript', () => {
+        describe('WHEN: javascript has been disabled', () => {
```
Comment (rmccar) [inline src/components/address-input/_macro.spec.js:146]:
```suggestion
        describe('WHEN: JavaScript has been disabled', () => {
```

**Review by balibirchlee — COMMENTED** [review]
This is a difficult one because I think the macro options file is a bit unwieldy - I think this one will need a pretty major revamp from the original test file. 
As there are a lot of required params - i would probably start with a "GIVEN: Params: required" and then do when/then's for each of those params
After that I would go through each macro option in order and do its given/when/then. if there's a group that makes sense to go together, then by all means group those, and just make sure it's obvious what params are being covered.

File: src/components/address-input/_macro.spec.js [hunk 23]
```diff
@@ -4,69 +4,22 @@ import * as cheerio from 'cheerio';
...
-    },
-};
-
-describe('macro: address-input', () => {
-    it('passes jest-axe checks', async () => {
-        const $ = cheerio.load(renderComponent('address-input', EXAMPLE_AUTOSUGGEST_ADDRESS_MINIMAL));
-
-        const results = await axe($.html());
-        expect(results).toHaveNoViolations();
+describe('FOR: address-input', () => {
```
Comment (rmccar) [inline src/components/address-input/_macro.spec.js:9]:
```suggestion
describe('FOR: Macro: Address-input', () => {
```
Same comment as other PR

File: src/components/address-input/_macro.spec.js [hunk 24]
```diff
@@ -4,69 +4,22 @@ import * as cheerio from 'cheerio';
...
+    describe('GIVEN: Params: required', () => {
+        describe('WHEN: All required params are provided', () => {
```
Comment (rmccar) [inline src/components/address-input/_macro.spec.js:11]:
```suggestion
        describe('WHEN: all required params are provided', () => {
```

**Review by rmccar — CHANGES_REQUESTED** [review]

File: src/components/address-input/_macro.spec.js [hunk 25]
```diff
@@ -4,69 +4,22 @@ import * as cheerio from 'cheerio';
...
-
-describe('macro: address-input', () => {
-    it('passes jest-axe checks', async () => {
-        const $ = cheerio.load(renderComponent('address-input', EXAMPLE_AUTOSUGGEST_ADDRESS_MINIMAL));
-
-        const results = await axe($.html());
-        expect(results).toHaveNoViolations();
+describe('FOR: Macro: Address-input', () => {
+    describe('GIVEN: Params: required', () => {
+        describe('WHEN: all required params are provided', () => {
```
Comment (rmccar) [inline src/components/address-input/_macro.spec.js:11]:
```suggestion
        describe('WHEN: required params are provided', () => {
```

File: src/components/address-input/_macro.spec.js [hunk 26]
```diff
@@ -87,126 +45,209 @@ describe('macro: address-input', () => {
...
-                id: 'address-input-example-id-line2',
-                name: 'address-input-example-id-line2',
-                classes: 'ons-js-address-line2',
-                label: {
-                    text: 'Address line 2',
-                },
-                value: '12345 The Road',
-                width: '20@m',
-                error: { text: 'Server error: address line 2' },
+            test('THEN: it renders autosuggest component with expected parameters', () => {
```
Comment (rmccar) [inline src/components/address-input/_macro.spec.js:108]: Maybe this should be "THEN: the provided attributes are passed through to the autosuggest component" because that is what it is testing

**Review by rmccar — CHANGES_REQUESTED** [review]
Couple of minor things

File: src/components/address-input/_macro.spec.js [hunk 27]
```diff
@@ -417,13 +346,15 @@ describe('macro: address-input', () => {
                 }),
             );
 
-            expect($('.ons-js-address-manual-btn').attr('href')).toBe('https://example.com/edit-address');
-            expect($('.ons-js-address-manual-btn').text().trim()).toBe('Manually enter address');
+            test('THEN: it renders the manualLink with the provided text and url', () => {
+                expect($('.ons-js-address-manual-btn').attr('href')).toBe('https://example.com/edit-address');
+                expect($('.ons-js-address-manual-btn').text().trim()).toBe('Manually enter address');
```
Comment (rmccar) [inline src/components/address-input/_macro.spec.js:351]: These should be split into individual thens

File: src/components/address-input/_macro.spec.js [hunk 28]
```diff
@@ -404,11 +331,13 @@ describe('macro: address-input', () => {
                 }),
             );
 
-            expect($('.ons-js-address-manual-btn').attr('href')).toBe('#0');
-            expect($('.ons-js-address-manual-btn').text().trim()).toBe('Manually enter address');
+            test('THEN: it renders the manual link with the provided text and a default url', () => {
+                expect($('.ons-js-address-manual-btn').attr('href')).toBe('#0');
+                expect($('.ons-js-address-manual-btn').text().trim()).toBe('Manually enter address');
```
Comment (rmccar) [inline src/components/address-input/_macro.spec.js:336]: These should be individual thens

File: src/components/address-input/_macro.spec.js [hunk 29]
```diff
@@ -404,11 +331,13 @@ describe('macro: address-input', () => {
...
+            });
         });
 
-        it('renders `manualLinkText` with `manualLinkUrl` when provided', () => {
+        describe('WHEN: a value for manualLinkUrl and manualLinkText is provided', () => {
```
Comment (rmccar) [inline src/components/address-input/_macro.spec.js:340]: Im not sure we need this test, its pretty much a duplicate of the one above

File: src/components/address-input/_macro.spec.js [hunk 30]
```diff
@@ -87,126 +45,209 @@ describe('macro: address-input', () => {
...
-                id: 'address-input-example-id-postcode',
-                name: 'address-input-example-id-postcode',
-                classes: 'ons-js-address-postcode',
-                label: {
-                    text: 'Postcode',
-                },
-                value: 'PO57 6ODE',
-                width: '7',
-                error: { text: 'Server error: postcode' },
+        describe('WHEN: Params: postcode', () => {
```
Comment (rmccar) [inline src/components/address-input/_macro.spec.js:232]: For these whens here they should be worded differently as they are worded more like a given. For example this one would be something like "WHEN: postcode is provided" and its the same with all the other parts of the address

**Review by rmccar — APPROVED** [review]

**Review by SriHV — APPROVED** [review]
